vm/src/builtins/enumerate.rs

```diff
@@ -76,32 +82,90 @@ impl PyValue for PyReverseSequenceIterator {
-    fn length_hint(&self) -> PyResult<isize> {
-        Ok(self.position.load() + 1)
+    fn length_hint(&self, vm: &VirtualMachine) -> PyResult<usize> {
+        Ok(match self.status.load() {
+            Active => {
+                let position = self.position.load();
+                if position > vm.obj_len(&self.obj)? {
+                    0
+                } else {
+                    position + 1
+                }
+            }
+            Exhausted => 0,
+        })
+    }
+
+    #[pymethod(magic)]
+    fn setstate(&self, state: PyObjectRef, vm: &VirtualMachine) -> PyResult<()> {
+        // When we're exhausted, just return.
+        if let Exhausted = self.status.load() {
+            return Ok(());
+        }
+        let len = vm.obj_len(&self.obj)?;
+        let pos = state
+            .payload::<PyInt>()
+            .ok_or_else(|| vm.new_type_error("an integer is required.".to_owned()))?;
+        let pos = std::cmp::min(
+            try_to_primitive(pos.as_bigint(), vm).unwrap_or(0),
+            len.saturating_sub(1),
+        );
+        self.position.store(pos);
+        Ok(())
+    }
+
+    #[pymethod(magic)]
+    fn reduce(&self, vm: &VirtualMachine) -> PyResult {
+        let iter = vm.get_attribute(vm.builtins.clone(), "reversed")?;
+        Ok(vm.ctx.new_tuple(match self.status.load() {
+            Exhausted => {
+                let obj_type = self.obj.clone_class();
+                let empty_obj = if let Some(ref tp_new) = obj_type.slots.new {
+                    tp_new(
+                        vm,
+                        FuncArgs {
+                            args: vec![obj_type.clone().into_object()],
+                            kwargs: IndexMap::new(),
+                        },
+                    )?
+                } else {
+                    vm.ctx.new_list(vec![])
+                };
+                vec![iter, vm.ctx.new_tuple(vec![empty_obj])]
+            }
+            Active => vec![
+                iter,
+                vm.ctx.new_tuple(vec![self.obj.clone()]),
+                vm.ctx.new_int(self.position.load()),
+            ],
+        }))
     }
 }
 
 impl PyIter for PyReverseSequenceIterator {
     fn next(zelf: &PyRef<Self>, vm: &VirtualMachine) -> PyResult {
+        if let Exhausted = zelf.status.load() {
+            return Err(vm.new_stop_iteration());
+        }
         let pos = zelf.position.fetch_sub(1);
-        if pos >= 0 {
-            match zelf.obj.get_item(pos, vm) {
-                Err(ref e) if e.isinstance(&vm.ctx.exceptions.index_error) => {
-                    Err(vm.new_stop_iteration())
-                }
-                // also catches stop_iteration => stop_iteration
-                ret => ret,
+        if pos == 0 {
+            zelf.status.store(Exhausted);
+        }
+        match zelf.obj.get_item(pos, vm) {
+            Err(ref e) if e.isinstance(&vm.ctx.exceptions.index_error) => {
+                Err(vm.new_stop_iteration())
```

Comment:
set the status here too. If it returns `StopIteration` once it should continue doing so.

Reply:
Thanks. I missed it. I applied it in v2 code 😊